src/main/utility/sockaddr.rs

```diff
@@ -0,0 +1,688 @@
+    /// returns the ipv6 socket address.
+    pub fn as_inet6_mut(&mut self) -> Option<&mut nix::sys::socket::SockaddrIn6> {
+        if !self.verify_fixed_size::<libc::sockaddr_in6>(AddressFamily::Inet6) {
+            return None;
+        }
+        // SAFETY: `nix::sys::socket::SockaddrIn6` should be a transparent wrapper around a
+        // `libc::sockaddr_in6`
+        Some(unsafe {
+            &mut *(&mut self.addr.inet6 as *mut _ as *mut nix::sys::socket::SockaddrIn6)
+        })
+    }
+
+    /// Get a new `SockaddrStorage` with a copy of the ipv6 socket address.
+    pub fn from_inet6(addr: &nix::sys::socket::SockaddrIn6) -> Self {
+        // SAFETY: `nix::sys::socket::SockaddrIn6` should be a transparent wrapper around a
+        // `libc::sockaddr_in6`
+        unsafe { Self::from_ptr(addr.as_ptr() as *const MaybeUninit<u8>, addr.len()) }.unwrap()
+    }
+
+    /// If the socket address represents a valid unix socket address (correct family and length),
+    /// returns the unix socket address.
+    pub fn as_unix(&self) -> Option<SockaddrUnix<&libc::sockaddr_un>> {
+        if self.family() != Some(AddressFamily::Unix) {
+            return None;
+        }
+
+        SockaddrUnix::new(unsafe { &self.addr.unix }, self.len)
+    }
+
+    /// If the socket address represents a valid unix socket address (correct family and length),
+    /// returns the unix socket address.
+    pub fn as_unix_mut(&mut self) -> Option<SockaddrUnix<&mut libc::sockaddr_un>> {
+        if self.family() != Some(AddressFamily::Unix) {
+            return None;
+        }
+
+        SockaddrUnix::new(unsafe { &mut self.addr.unix }, self.len)
+    }
+
+    /// Get a new `SockaddrStorage` with a copy of the unix socket address.
+    pub fn from_unix(addr: &SockaddrUnix<&libc::sockaddr_un>) -> Self {
+        let (ptr, len) = addr.as_ptr();
+
+        unsafe { Self::from_ptr(ptr as *const MaybeUninit<u8>, len) }.unwrap()
+    }
+
+    /// A pointer to the socket address. Some bytes may be uninitialized.
+    pub fn as_ptr(&self) -> (*const MaybeUninit<u8>, libc::socklen_t) {
+        (unsafe { &self.addr.slice }.as_ptr(), self.len)
+    }
+
+    /// The socket address as a slice of bytes. Some bytes may be uninitialized.
+    pub fn as_slice(&self) -> &[MaybeUninit<u8>] {
+        unsafe { &self.addr.slice[..(self.len as usize)] }
+    }
+
+    /// Returns true if the address length is large enough for a `AsType`, and if it has the
+    /// provided family.
+    fn verify_fixed_size<AsType>(&self, family: AddressFamily) -> bool {
+        (self.len as usize) >= std::mem::size_of::<AsType>() && self.family() == Some(family)
+    }
+}
+
+impl std::fmt::Debug for SockaddrStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SockaddrStorage")
+            .field("family", &self.family())
+            .field("len", &self.len)
+            .finish_non_exhaustive()
```

Comment:
Optional - if we're ensuring all bytes up to len are initialized, I think we could include the raw bytes here? Or if we're ensuring that the family byte is correct, maybe we could pretty-print the actual address?

Reply:
I expanded this to try casting as the different address types first.